### ISSUE TITLE
Await bloc.add

### ISSUE THREAD

**Issue by ollydixon** [felangel/bloc#1556]
For many reasons it would be awesome if we could wait for the bloc to finish with the event before continuing.

Here's an example

```
Future<void> testBloc(
) async {
  final data = await request('fetch-user');
  final user = UserV1.fromJson(data);
  // Bloc is access via static reference, it has to be this way.
  userBloc.add(UpdateUserAction(user));
  
  // Fails because it needs userBloc.state.user.uid
  fetchAllLogs();
}
```

I can't seem to find a solution or feature for it. If anyone knows it would be great.

**Comment by RollyPeres** [comment]
Hi @ollydixon 👋

Why not invoke `fetchAllLogs` in the `UpdateUserAction` handler ? e.g.:

```
if (event is UpdateUserAction) {
  // process event
  fetchAllLogs();
}
```

**Comment by ollydixon** [comment]
@RollyPeres fetchAllLogs also has a logsBloc with add. :-/ I'm sort of stuck here; thanks for the reply.

**Comment by felangel** [comment]
Hi @ollydixon 👋 
Thanks for opening an issue!

If you really need to do this you can do

```dart
userBloc.add(UpdateUserAction(user));

final nextState = await userBloc.first;

fetchAllLogs();
```

But this is usually not the best solution and is likely an indication that the flow of data can be reworked/simplified.

**Comment by ollydixon** [comment]
@felangel thank you so much! I've been stuck on this all day.

**Comment by ollydixon** [comment]
@felangel it would seem that the state is still invalid even after the await.

For example.

`await userBloc.first;`
`userBloc.state.logs` << 0

But if I wait for a second then run it again, userBloc.state.logs is 151.

Hmm stumped.

**Comment by felangel** [comment]
await `userBloc.first` would wait for the next state. If you're yielding multiple states for a single event you need to do `firstWhere`

**Comment by ollydixon** [comment]
@felangel I'm guessing firstWhere will just return the next state that was just added? (Matching the bool)

**Comment by felangel** [comment]
@ollydixon yup 👍

**Comment by felangel** [comment]
Closing this for now but feel free to comment with additional questions and I'm happy to continue the conversation 👍

**Comment by ollydixon** [comment]
@felangel thanks :-)

Would you ever consider adding await bloc.add to guarantee a consistent API for say 'next' event processes. Just an idea; maybe I sound dumb.

Great job on this lib btw. We deleted all our Redux from over 30 Flutter projects at my company.

**Comment by felangel** [comment]
@ollydixon I would like to avoid that because `add` is part of the `Sink` api which does not return a `Future` and also it would be confusing because a single add can result in multiple state changes so you would never know when the event was actually "done" being processed.

**Comment by ollydixon** [comment]
@felangel makes sense. I guess it's never reliable to assume values from distant state anyhow. Sticking to the firstWhere seemed to work perfect for our auth fetching scenarios. :-)

**Comment by felangel** [comment]
You can accomplish this more easily using `Cubit`

```dart
class MyCubit extends Cubit<MyState> {
  MyCubit() : super(MyState.initial());

  Future<void> doSomething() async {
    emit(MyState.loading());
    final result = await _repository.getResult();
    emit(MyState.success(result));
  }
}
```

Then you can await `doSomething` directly:

```dart
final cubit = MyCubit();
await cubit.doSomething();
print(cubit.state); // MyState.success
```

Hope that helps and thanks so much for the positive feedback -- I really appreciate it 🙏

**Comment by ollydixon** [comment]
Thanks, I'll look into this; little abstract for my brain haha; most of our state/actions are in order, the stream stuff always confuses my coders.

For example it's as simple as:

```
Future<UserV1> updateUser(Map<String, dynamic> data) async {
  try {
    final json = await request('update-user', data: data);
    final user = UserV1.fromJson(json);
    userBloc.add(UpdateUser(user));
    return updatedUser;
  } catch (e) {
    print('Error updating user');
    print(e);
    throw e;
  }
}
```

The the UI handles the rest. Errors are handled directly and not in state management. We like as little code as possible! :-)

**Comment by ericderegt** [comment]
I was using firstWhere for a similar use case and it was working fine. When I upgraded to 7.0.0 I get a message saying that firstWhere isn't defined. Is this the expected behavior? I couldn't find anything in the release notes? Here's my code

```
                      _loginBloc.add(SignInWithPhoneEvent(
                          verificationCode: _verificationCode,
                          phone: widget.user.phone!,
                          pin: pin));

                      // https://github.com/felangel/bloc/issues/1556
                      final nextState = await _loginBloc.firstWhere((element) =>
                          element is LoginNeedsRegister ||
                          element is LoginSuccess);
```

**Comment by yfer** [comment]
@ericderegt in 7.0 you should use loginBloc.stream.firstWhere instead of loginBloc.firstWhere

**Comment by minhaeng** [comment]
I solved this problem just using callback after emit states. In your case, just send event with fetchAllLogs and call it after emitting states. In this way, i solved problem of execution order.
